src/Microsoft.DotNet.TestFramework/TestAssetInstance.cs

```diff
@@ -156,19 +174,141 @@ public TestAssetInstance UseCurrentRuntimeFrameworkVersion()
             });
         }
 
-        private void CopyFiles(IEnumerable<FileInfo> filesToCopy)
+        private static string RebasePath(string path, string oldBaseDirectory, string newBaseDirectory)
+        {
+            path = Path.IsPathRooted(path) ? PathUtility.GetRelativePath(PathUtility.EnsureTrailingSlash(oldBaseDirectory), path) : path;
+            return Path.Combine(newBaseDirectory, path);
+        }
+
+        private void CopySourceFiles()
         {
+            var filesToCopy = TestAssetInfo.GetSourceFiles();
             foreach (var file in filesToCopy)
             {
-                var relativePath = file.FullName.Substring(TestAssetInfo.Root.FullName.Length + 1);
-
-                var newPath = Path.Combine(Root.FullName, relativePath);
+                var newPath = RebasePath(file.FullName, TestAssetInfo.Root.FullName, Root.FullName);
 
                 var newFile = new FileInfo(newPath);
 
                 PathUtility.EnsureDirectoryExists(newFile.Directory.FullName);
 
-                file.CopyTo(newPath);
+                CopyFileAdjustingPaths(file, newFile);
+            }
+        }
+
+        private bool IsAncestor(FileInfo file, DirectoryInfo maybeAncestor)
+        {
+            var dir = file.Directory;
+            do
+            {
+                if (string.Equals(maybeAncestor.FullName, dir.FullName, StringComparison.OrdinalIgnoreCase))
```

Comment:
@piotrpMSFT Can you take a look here? You mentioned we may have an extension for this already.

Also, @krwq if you ran into problems during your tests or CI, I can understand writing a more complicated approach, but otherwise, I prefer a simpler solution until the more complicated one is needed.